Excessive MPI_Handler_function deprecated warnings: just from including mpi.h

We recently added `--enable-mpi1-compatibility` and by default no longer compile things that were removed from the standard long ago.  We also added deprecation markers to scold users when they use these ancient functions.  These are generally Good Things.

However, if you compile with `--enable-mpi1-compatibilty`, just the act of `#include <mpi.h>` results in deprecation warnings.  For example, here's compiling `examples/hello_c.c`:

```
mpicc -g  hello_c.c  -o hello_c
In file included from hello_c.c:11:0:
/home/jsquyres/bogus/include/mpi.h:2673:1: warning: ‘MPI_Handler_function’ is deprecated: MPI_Handler_function was removed in MPI-3.0; use MPI_Win_errhandler_function instead [-Wdeprecated-declarations]
 OMPI_DECLSPEC  int MPI_Errhandler_create(MPI_Handler_function *function,
 ^
/home/jsquyres/bogus/include/mpi.h:2712:1: warning: ‘MPI_Handler_function’ is deprecated: MPI_Handler_function was removed in MPI-3.0; use MPI_Win_errhandler_function instead [-Wdeprecated-declarations]
 OMPI_DECLSPEC  int PMPI_Errhandler_create(MPI_Handler_function *function,
 ^
```

This is because `MPI_Handler_function` is a deprecated/removed type, but it is used in the MPI-1 functions `[P]MPI_Errhandler_create()`, which are prototyped in `mpi.h`.

We should not be causing warnings like this *just for enabling MPI-1 compatibility* -- specifically when the MPI application is not using any MPI-1 functionality.

What is odd, however, is that we don't get warnings for other function typedefs, like `MPI_Copy_function`, and others.  This is what I don't understand, and why I haven't just created a PR that simply removes the deprecation warning from `MPI_Handler_function` in `mpi.h`.

Does anyone know why this is happening?

FYI @hjelmn

--- a/ompi/include/mpi.h.in
+++ b/ompi/include/mpi.h.in
@@ -9,7 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
- * Copyright (c) 2007-2016 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2007-2018 Cisco Systems, Inc.  All rights reserved
  * Copyright (c) 2008-2009 Sun Microsystems, Inc.  All rights reserved.
  * Copyright (c) 2009-2012 Oak Rigde National Laboratory.  All rights reserved.
  * Copyright (c) 2011      Sandia National Laboratories. All rights reserved.
@@ -419,17 +419,21 @@ typedef int MPIX_Grequest_class;
 /*
  * Removed typedefs. These will be deleted in a future Open MPI release.
  */
-typedef int (MPI_Copy_function)(MPI_Comm, int, void *,
-                                void *, void *, int *);
-typedef int (MPI_Delete_function)(MPI_Comm, int, void *, void *);
 typedef MPI_Comm_errhandler_function MPI_Comm_errhandler_fn
         __mpi_interface_removed__("MPI_Comm_errhandler_fn was removed in MPI-3.0; use MPI_Comm_errhandler_function instead");
 typedef ompi_file_errhandler_fn MPI_File_errhandler_fn
         __mpi_interface_removed__("MPI_File_errhandler_fn was removed in MPI-3.0; use MPI_File_errhandler_function instead");
 typedef MPI_Win_errhandler_function MPI_Win_errhandler_fn
         __mpi_interface_removed__("MPI_Win_errhandler_fn was removed in MPI-3.0; use MPI_Win_errhandler_function instead");
-typedef void (MPI_Handler_function)(MPI_Comm *, int *, ...)
-        __mpi_interface_removed__("MPI_Handler_function was removed in MPI-3.0; use MPI_Win_errhandler_function instead");
+
+// NOTE: We intentionally do *not* mark the following as
+// deprecated/removed because they are used below in function
+// prototypes (and would therefore emit warnings, just for #including
+// <mpi.h>).
+typedef void (MPI_Handler_function)(MPI_Comm *, int *, ...);
+typedef int (MPI_Copy_function)(MPI_Comm, int, void *,
+                                void *, void *, int *);
+typedef int (MPI_Delete_function)(MPI_Comm, int, void *, void *);
 #endif /* !OMPI_OMIT_MPI1_COMPAT_DECLS */
 
 

